packages/react-accordion/src/components/AccordionHeader/useAccordionHeader.ts

```diff
@@ -35,58 +35,49 @@ export const accordionHeaderShorthandProps: Array<keyof AccordionHeaderSlots> =
  */
 export const useAccordionHeader = (props: AccordionHeaderProps, ref: React.Ref<HTMLElement>): AccordionHeaderState => {
   const { onHeaderClick: onAccordionHeaderClick, disabled, open } = useAccordionItemContext();
-  const button = useContextSelector(AccordionContext, ctx => ctx.button);
-  const expandIcon = useContextSelector(AccordionContext, ctx => ctx.expandIcon);
-  const inline = useContextSelector(AccordionContext, ctx => ctx.inline);
-  const icon = useContextSelector(AccordionContext, ctx => ctx.icon);
-  const expandIconPosition = useContextSelector(AccordionContext, ctx => ctx.expandIconPosition);
-  const size = useContextSelector(AccordionContext, ctx => ctx.size);
   const id = useId('accordion-header-', props.id);
   const panel = useDescendants(accordionItemDescendantContext)[1] as AccordionItemDescendant | undefined;
   const innerRef = React.useRef<HTMLElement>(null);
-  const initialState = {
+
+  useAccordionItemDescendant(
+    {
+      element: innerRef.current,
+      id,
+    },
+    0,
+  );
+
+  const originalHandleClick = isObjectShorthandProps(props.button) ? props.button.onClick : undefined;
+
+  return {
     ...props,
+    disabled,
+    open,
     ref: useMergedRefs(ref, innerRef),
-    size: props.size ?? size ?? 'medium',
-    inline: props.inline ?? inline ?? false,
-    role: props.role ?? 'heading',
-    expandIconPosition: props.expandIconPosition ?? expandIconPosition ?? 'start',
+    size: props.size || 'medium',
+    inline: props.inline || false,
+    role: props.role || 'heading',
+    expandIconPosition: props.expandIconPosition || 'start',
     components: {
       root: 'div',
       button: 'button',
       expandIcon: AccordionHeaderExpandIcon,
     },
-    icon: resolveShorthand(props.icon ?? icon),
-    expandIcon: resolveShorthand(props.expandIcon ?? expandIcon, {
+    icon: resolveShorthand(props.icon),
+    expandIcon: resolveShorthand(props.expandIcon, {
       'aria-hidden': true,
     }),
-    button: useARIAButton(props.button ?? button, {
+    button: useARIAButton(props.button, {
       id,
-      onClick: onAccordionHeaderClick,
+      onClick: useEventCallback((ev: React.MouseEvent<HTMLElement>) => {
+        originalHandleClick?.(ev);
+        if (!ev.isDefaultPrevented()) {
+          onAccordionHeaderClick(ev);
+        }
+      }),
```

Comment:
**TL;DR** this should be an override on `state`, i.e. `state.button.onClick = useEventCallback()`

---

To verify I added a log into `useEventCallback()` and modifier `AccordionItem.stories.tsx`:

```diff
    button: useARIAButton(props.button, {
      id,
      onClick: useEventCallback((ev: React.MouseEvent<HTMLElement>) => {
+       console.log('useAccordionHeader');
        originalHandleClick?.(ev);
        if (!ev.isDefaultPrevented()) {
          onAccordionHeaderClick(ev);
        }
      }),
```

```diff
        <AccordionHeader
+         button={{ onClick: () => console.log('button.onClick') }}
          inline={args.inline}
          icon={args.icon ? <RocketIcon /> : undefined}
```

When I click on header I am getting only one log:

![image](https://user-images.githubusercontent.com/14183168/125053550-89241100-e0a5-11eb-8f27-0789f1180c92.png)

With proper implementation I should have both:

![image](https://user-images.githubusercontent.com/14183168/125053724-bcff3680-e0a5-11eb-9408-9ff855b8bb8e.png)

---

**Why it happens?** `useARIAButton()` uses `resolveShorthand()` that as the second argument has `defaultProps` that has lower priority than `props` in first param:

https://github.com/microsoft/fluentui/blob/394b16b92fae7773125c32edc1df56c7758cba4e/packages/react-utilities/src/compose/resolveShorthand.ts#L22

**How to fix it?** As I mentioned, an override will fix an issue there. From wider perspective, I think that we should modify `defaultProps` typings to accept only primitive values (`null`, `boolean`, `number`, `string`) as only they could be merged without issues with `Object.assign()`.

---

Optionally, we can implement the same thing as in SUIR/Fluent N*:

```tsx
    button: useARIAButton(props.button, {
      defaultProps: { id },
      overrideProps: (predefinedProps) => ({
        onClick: useEventCallback((ev: React.MouseEvent<HTMLElement>) => {
          predefinedProps?.onClick(ev);
        }),
     })
   })
```

But I don't think that it's a good idea.

---

P.S. I also suggest a unit test for this case.

Reply:
Nice, I'll add a UT immediately and think about a possible good solution